### PR TITLE
[X86][MC] Support Enc/Dec for EGPR for promoted CET instruction

### PR DESCRIPTION
**Pull request by XinWang10**
R16-R31 was added into GPRs in https://github.com/llvm/llvm-project/pull/70958,
This patch supports the encoding/decoding for promoted CET instruction in EVEX space.

RFC: https://discourse.llvm.org/t/rfc-design-for-apx-feature-egpr-and-ndd-support/73031/4

### CODE REVIEW

File: llvm/test/MC/Disassembler/X86/apx/wrssq.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# RUN: llvm-mc --disassemble %s -triple=x86_64 | FileCheck %s --check-prefixes=ATT
+# RUN: llvm-mc --disassemble %s -triple=x86_64 -x86-asm-syntax=intel --output-asm-variant=1 | FileCheck %s --check-prefixes=INTEL
```
Comment (KanRobert) [inline llvm/test/MC/Disassembler/X86/apx/wrssq.txt:2]: Could you merge the test  WRUSSD with WRUSSQ? `D`, `Q` here is kind of a suffix here.

Comment (XinWang10) [reply llvm/test/MC/Disassembler/X86/apx/wrssq.txt:2]: They are listed as different instructions in document.

File: llvm/lib/Target/X86/X86InstrSystem.td [hunk 2]
```diff
@@ -532,6 +533,22 @@ let SchedRW = [WriteSystem] in {
...
+  def WRSSQ_EVEX : RI<0x66, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
+                      "wrssq\t{$src, $dst|$dst, $src}",
+                      [(int_x86_wrssq GR64:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PS;
+  def WRUSSD_EVEX : I<0x65, MRMDestMem, (outs), (ins i32mem:$dst, GR32:$src),
+                      "wrussd\t{$src, $dst|$dst, $src}",
+                      [(int_x86_wrussd GR32:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PD;
+  def WRUSSQ_EVEX : RI<0x65, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
+                       "wrussq\t{$src, $dst|$dst, $src}",
+                       [(int_x86_wrussq GR64:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PD;
+}
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrSystem.td:551]: Do we need to add IR test for the pattens, or leave them blank if we only cover encoding/decosing part?

File: llvm/lib/Target/X86/X86InstrSystem.td [hunk 3]
```diff
@@ -520,6 +520,7 @@ let SchedRW = [WriteSystem] in {
     } // Defs SSP
   } // Uses SSP
 
+let Predicates = [NoEGPR] in {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrSystem.td:523]: How about other CET instuctions like `RDSSPD`? It has GPR operand too. It looks to me strange the symmetry instuction doesn't gain APX promotion.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrSystem.td:523]: `RDSSPD` is not promoted. I guess it's probably b/c it's used by intrinsic and the dest reg is `RAX` in most cases.

File: llvm/lib/Target/X86/X86InstrSystem.td (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrSystem.td:551]: We also miss the IR test for previous promoted instructions. I believe @XinWang10 will add them when he has bandwidth.

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSystem.td:551]: Yes, IR test would be added later, and if may not affect test with out egpr feature.

File: llvm/lib/Target/X86/X86InstrSystem.td [hunk 4]
```diff
@@ -532,6 +533,22 @@ let SchedRW = [WriteSystem] in {
...
+                     [(int_x86_wrssd GR32:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PS;
+  def WRSSQ_EVEX : RI<0x66, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
+                      "wrssq\t{$src, $dst|$dst, $src}",
+                      [(int_x86_wrssq GR64:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PS;
+  def WRUSSD_EVEX : I<0x65, MRMDestMem, (outs), (ins i32mem:$dst, GR32:$src),
+                      "wrussd\t{$src, $dst|$dst, $src}",
+                      [(int_x86_wrussd GR32:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PD;
+  def WRUSSQ_EVEX : RI<0x65, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
+                       "wrussq\t{$src, $dst|$dst, $src}",
+                       [(int_x86_wrussq GR64:$src, addr:$dst)]>, EVEX_NoCD8, T_MAP4PD;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrSystem.td:550]: EVEX_NoCD8 -> EVEX, NoCD8

EVEX_NoCD8 was removed for the coming NDD support.

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSystem.td:550]: Will rebase

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**